### ISSUE TITLE
some question about mxnet gpu memory usage

### ISSUE THREAD

**Issue by chenboheng** [apache/incubator-mxnet#3095]
Many people say that mxnet is a memory efficient framework.
And I test the caltech101 classification experiment by train_imagenet.py and symbol_alexnet.py.
I convert the images to train.lst and test.lst by im2rec.py. Then I convert train.lst and test.lst to train.rec and test.rec (The number of training images, testing images and class are 3657, 5486 and 102 respectively) by im2rec. At last, I use the alexnet to do this classification experiment. 
## The train_imagenet.py code is:

parser = argparse.ArgumentParser(description='train an image classifer on imagenet')
parser.add_argument('--network', type=str, default='alexnet',
                    help = 'the cnn to use')
parser.add_argument('--data-dir', type=str, default='/media/ResearchProject/deeplearning/code/mxnet-experiments/data/caltech101/',
                    help='the input data directory')
parser.add_argument('--model-prefix', type=str,
                    help='the prefix of the model to load')
parser.add_argument('--save-model-prefix', type=str,
                    help='the prefix of the model to save')
parser.add_argument('--lr', type=float, default=.001,
                    help='the initial learning rate')
parser.add_argument('--lr-factor', type=float, default=0.1,
                    help='times the lr with a factor for every lr-factor-epoch epoch')
parser.add_argument('--lr-factor-epoch', type=float, default=30,
                    help='the number of epoch to factor the lr, could be .5')
parser.add_argument('--clip-gradient', type=float, default=5.,
                    help='clip min/max gradient to prevent extreme value')
parser.add_argument('--num-epochs', type=int, default=120,
                    help='the number of training epochs')
parser.add_argument('--load-epoch', type=int,
                    help="load the model on an epoch using the model-prefix")
parser.add_argument('--batch-size', type=int, default=64,
                    help='the batch size')
parser.add_argument('--gpus', type=str,default='0',
                    help='the gpus will be used, e.g "0,1,2,3"')
parser.add_argument('--kv-store', type=str, default='local',
                    help='the kvstore type')
parser.add_argument('--num-examples', type=int, default=3657,
                    help='the number of training examples')
parser.add_argument('--num-classes', type=int, default=102,
                    help='the number of classes')
parser.add_argument('--log-file', type=str, 
            help='the name of log file')
parser.add_argument('--log-dir', type=str, default="/tmp/",
                    help='directory of the log file')
parser.add_argument('--train-dataset', type=str, default="train.rec",
                    help='train dataset name')
parser.add_argument('--val-dataset', type=str, default="test.rec",
                    help="validation dataset name")
parser.add_argument('--data-shape', type=int, default=256,
                    help='set image\'s shape')
args = parser.parse_args()
# network

import importlib
net = importlib.import_module('symbol_' + args.network).get_symbol(args.num_classes)
# data

def get_iterator(args, kv):
    data_shape = (3, args.data_shape, args.data_shape)
    train = mx.io.ImageRecordIter(
        path_imgrec = os.path.join(args.data_dir, args.train_dataset),
        mean_img = args.data_dir+'mean.bin',
        data_shape  = data_shape,
        batch_size  = args.batch_size,
        rand_crop   = True,
        rand_mirror = True,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)

```
val = mx.io.ImageRecordIter(
    path_imgrec = os.path.join(args.data_dir, args.val_dataset),
    mean_img=args.data_dir + 'mean.bin',
    rand_crop   = False,
    rand_mirror = False,
    data_shape  = data_shape,
    batch_size  = args.batch_size,
    num_parts   = kv.num_workers,
    part_index  = kv.rank)

return (train, val)
```
# train
## train_model.fit(args, net, get_iterator)

But I found that the gpu memory usage of mxnet is apparently higer than caffe.

The memory usage of mxnet:

(1) gpu doesn't work:
![image](https://cloud.githubusercontent.com/assets/20148944/17844944/f0d38388-6870-11e6-8bcb-8ce992061ba2.png)

(2) batch_size=1
![image](https://cloud.githubusercontent.com/assets/20148944/17844965/0cc6110a-6871-11e6-9523-c481c66a931b.png)

(3) batch_size=64
![image](https://cloud.githubusercontent.com/assets/20148944/17844986/36228718-6871-11e6-9168-136518255c21.png)

(4)batch_size=128
![image](https://cloud.githubusercontent.com/assets/20148944/17844997/4cc0a0e0-6871-11e6-949c-3cabc67d6183.png)

And when using caffenet(Alexnet) in caffe, the training batch_size and test batch_size are set as 64, the gpu memory usage of caffe is 2463MB.

mxnet uses higher memory than caffe in the same Alexnet structure. These things confuse me. So I want to know whether I am wrong or the mxnet is not memory efficient. My configuration is ubuntu14.04 cuda8.0 gtx1080 cudnn5.1

**Comment by piiswrong** [comment]
1. use ccsgd optimizer instead of sgd
2. system memory doesn't reflect real usage of memory due to delayed release. Increase batch_size until OOM and compare the maximum viable batch_size to properly compare memory usage

**Comment by chenboheng** [comment]
Thank you for your advice, and how to use ccsgd optimizer? Thank you very much.

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
